### PR TITLE
STM32 QSPI Flash support

### PR DESCRIPTION
**Pull request by dgarske**
* STM32 QSPI Flash support.
* Refactor SPI to allow different GPIO base/AF for each pin.

Tested on STM32H753ZI hardware.

### CODE REVIEW

**Review by tatowicz — APPROVED** [review]
Small comment looks good otherwise

File: test-app/app_stm32h7.c [hunk 1]
```diff
@@ -259,14 +268,23 @@ int uart_setup(uint32_t bitrate)
     GPIOD_MODE = reg | (2 << (UART_RX_PIN * 2));
 
     /* Alternate function. Use AFLR for pins 0-7 and AFHR for pins 8-15 */
+#if UART_TX_PIN >= 8
+    reg = GPIOD_AFRH & ~(0xf << ((UART_TX_PIN & 0x07)*4));
```
[the remark this answers is not in the document]
Comment (dgarske) [reply test-app/app_stm32h7.c:272]: No

**Review by danielinux — APPROVED** [review]

**Comment by danielinux** [comment]
:x: 

SPI jenkins test fails on this branch
(`test-21-forward-update-no-downgrade-SPI` does not boot when SPI is on)

:x:

**Review by danielinux — CHANGES_REQUESTED** [review]
There is an issue in stm_gpio_config. 

By fixing this, spi test 21 boots, but fails to complete the update afterwards.

File: hal/spi/spi_drv_stm32.c [hunk 2]
```diff
@@ -25,146 +25,239 @@
...
-#if SPI1_MISO_PIN < 8
-    reg = SPI_PIO_AFL & ~(0xf << ((SPI1_MISO_PIN) * 4));
-    SPI_PIO_AFL = reg | (SPI1_PIN_AF << ((SPI1_MISO_PIN) * 4));
-#else
-    reg = SPI_PIO_AFH & ~(0xf << ((SPI1_MISO_PIN-8) * 4));
-    SPI_PIO_AFH = reg | (SPI1_PIN_AF << ((SPI1_MISO_PIN - 8) * 4));
-#endif
-
-#ifdef PLATFORM_stm32l0
-    reg = SPI_PIO_PUPD & ~(0x03 <<  (SPI1_CLOCK_PIN * 2));
```
Comment (danielinux) [inline hal/spi/spi_drv_stm32.c:110]: Please add a `PLATFORM_stm32l0` exception in `stm_pins_setup` for stm32l0 to set the appropriate pulldown and ospd during SPI pin initialization

**Comment by danielinux** [comment]
Closing in favor of #264

**Pull request closed without merging**